Sources/core/objects/Gun.cpp

```diff
@@ -0,0 +1,96 @@
+#include "Gun.h"
+#include "Tile.h"
+
+Gun::Gun(size_t id) : Item(id)
+{
+    SetSprite("icons/guns.dmi");
+    ammunition_ = 1;
+    max_ammunition_ = 1;
+}
+
+bool Gun::UseAmmo()
+{
+    if (ammunition_ <= 0)
+    {
+        ammunition_ = 0;
+        return false;
+    }
+    --ammunition_;
+    return true;
+}
+
+void Gun::Shoot(VDir target)
+{
+    id_ptr_on<CubeTile> tile = GetOwner()->GetOwner();
```

Comment:
Also, you overall use a lot of copy-paste.
For example, the `Shoot` functions. `Gun::Shoot`, `Laser::Shoot` and `Revolver::Shoot` all use practically same code (only differences is the type of projectile). You can make it a lot shorter if you will make something like that
```
void Gun::ShootImpl(VDir target, const std::string& projectile_type /* also some other params if needed */)
{
    ...
    auto projectile = Create<Projectile>(projectile_type, tile);
    ...
}
void LasetGun::Shoot(VDir target)
{
    ShootImpl(target, Laser::T_ITEM_S());
}
```
It will make code much cleaner.